README.md
# request-replay [![Build Status](https://secure.travis-ci.org/godfat/request-replay.png?branch=master)](http://travis-ci.org/godfat/request-replay)

by Lin Jen-Shin ([godfat](http://godfat.org))

## LINKS:

* [github](https://github.com/godfat/request-replay)
* [rubygems](https://rubygems.org/gems/request-replay)
* [rdoc](http://rdoc.info/github/godfat/request-replay)

## DESCRIPTION:

Replay the request via Rack env

## REQUIREMENTS:

* Tested with MRI (official CRuby), Rubinius and JRuby.

## INSTALLATION:

    gem install request-replay

## SYNOPSIS:

You might want to use this as middleware to replay the request:

Note that the first argument is where it should make the request,
and the second argument is what's additional headers we want to
overwrite in the original request.

``` ruby
require 'request-replay'
use RequestReplay::Middleware, 'localhost:8080',
    :add_headers => {'Host' => 'example.com'},
    :read_wait   => 5,
    # We could also rewrite the env
    :rewrite_env => lambda{ |env|
                      if env['HTTP_HOST'].start_with?('api.')
                        env['PATH_INFO'] = "/api/#{env['PATH_INFO']}"
                      end
                      env
                    }
run lambda{ |env| [200, {}, [env.inspect]] }
```

You could also use `RequestReplay::Proxy` as a reverse proxy. Note that
this only works on Rack servers which support [Rack Hijacking][].

[Rack Hijacking]: http://rack.rubyforge.org/doc/SPEC.html

``` ruby
require 'request-replay'
run RequestReplay::Proxy.new(
  'example.com', :add_headers => {'Host' => 'example.com'},
                 # We could also rewrite the env
                 :rewrite_env => lambda{ |env|
                   if env['HTTP_HOST'].start_with?('api.')
                     env['PATH_INFO'] = "/api/#{env['PATH_INFO']}"
                   end
                   env
                 })
```

## CONTRIBUTORS:

* Jim Wang (@yyjim)
* Lin Jen-Shin (@godfat)

## LICENSE:

Apache License 2.0

Copyright (c) 2013-2014, Lin Jen-Shin (godfat)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

<http://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGES.md
# CHANGES

## request-replay 0.7.1 -- 2014-03-30

* Add a default header: `Connection: close` for RequestReplay::Proxy since
  we're not going to reuse the connection anyway. There's no way to return
  the socket back to the web server after hijacking according to Rack's SPEC.
  This should fix some weird issues with web pages with a lot of images.

## request-replay 0.7.0 -- 2014-03-30

* Fixed a bug where nginx with sendfile on might not send a full file back.
* Introduced RequestReplay::Proxy which could serve as a reverse proxy.

## request-replay 0.6.3 -- 2013-10-07

* Fixed an issue where Rack::Request does not try to rewind rack.input for
  form POST. We rewind for them. Thanks @yyjim
* Extracted more constants. Might boost performance a bit.

## request-replay 0.6.2 -- 2013-10-01

* Added :rewrite_env option for rewriting env for specific use.

## request-replay 0.6.1 -- 2013-10-01

* Print error messages to env['rack.errors']
* Fixed a bug where the underlying Rack app might be modifying
  the original env, making request-replay cannot reliably rebuild
  the request.

## request-replay 0.6.0 -- 2013-10-01

* Added :read_wait option for waiting for the remote server responding.
* Changed the API a bit. Now pass :add_headers for adding some extra
  headers to the replaying request. Useful for passing another Host or
  a special User-Agent.

## request-replay 0.5.0 -- 2013-09-04

* Birthday!


test/test_basic.rb

require 'bacon'
require 'muack'

Bacon.summary_on_exit
Bacon::Context.__send__(:include, Muack::API)

module Kernel
  def eq? rhs
    self == rhs
  end
end

require 'request-replay'
require 'rack'

describe RequestReplay do
  host = 'localhost'.freeze
  port = 1024 + rand(2**16 - 1024)
  serv = TCPServer.new('localhost', port)
  hopt = "#{host}:#{port}".freeze
  env  = {'REQUEST_METHOD' => 'GET',
          'PATH_INFO' => '/', 'QUERY_STRING' => 'q=1',
          'HTTP_HOST' => 'localhost',
          'HTTP_PORK' => 'BEEF'     }.freeze

  verify = lambda do |response, expected|
    sock = serv.accept
    sock.read     .should.eq(expected)
    sock.write(expected)
    sock.close
    response.value.should.eq(expected)
  end

  request = lambda do |env1, headers={}, read_wait=nil|
    Thread.new(RequestReplay.new(env.merge(env1), hopt,
                                 :add_headers => headers,
                                 :read_wait   => read_wait)) do |replay|
      replay.start(&:read)
    end
  end

  after do
    Muack.verify
  end

  should 'GET' do
    verify[request[{'REQUEST_METHOD' => 'GET'}, 'Host' => 'ex.com'], <<-HTTP]
GET /?q=1 HTTP/1.1\r
Host: ex.com\r
Pork: BEEF\r
\r
    HTTP
  end

  should 'POST' do
    verify[request['REQUEST_METHOD' => 'POST',
                   'QUERY_STRING'   => ''    , # test no query string
                   'PATH_INFO'      => ''    , # test no path info
                   'rack.input' => StringIO.new("PAYLOAD\r\n\r\n")], <<-HTTP]
POST / HTTP/1.1\r
Host: localhost\r
Pork: BEEF\r
\r
PAYLOAD\r
\r
    HTTP
  end

  should 'read_wait' do
    read_wait = 5
    mock(IO).select(satisfy{ |rs| rs.size == 1 &&
                                  rs[0].kind_of?(IO) },
                    [], [], read_wait)

    verify[request[{}, {}, read_wait], <<-HTTP]
GET /?q=1 HTTP/1.1\r
Host: localhost\r
Pork: BEEF\r
\r
    HTTP
  end

  should 'puts error' do
    any_instance_of(TCPSocket) do |sock|
      mock(sock).read{ raise 'ERROR' }
    end

    errors = StringIO.new
    begin
      request['rack.errors' => errors].value
    ensure
      serv.accept.close
    end
    errors.string.should.start_with? '[RequestReplay] Error:'
  end

  should 'not affect Rack::Request' do
    input = StringIO.new('a=0&b=1')
    e     = {'rack.input' => input, 'REQUEST_METHOD' => 'POST'}
    t     = request[e]
    serv.accept.close
    t.join
    Rack::Request.new(e).POST.should.eq('a' => '0', 'b' => '1')
  end

  describe RequestReplay::Middleware do
    should 'PUT' do
      app = Rack::Builder.app do
        use RequestReplay::Middleware, hopt
        run lambda{ |_| [200, {}, []] }
      end

      app.call(env.merge('REQUEST_METHOD' => 'PUT'))
      begin
        sock = serv.accept
        sock.read.should.eq <<-HTTP
PUT /?q=1 HTTP/1.1\r
Host: localhost\r
Pork: BEEF\r
\r
      HTTP
      ensure
        sock.close
      end
    end

    should 'retain original env' do
      e = env.dup

      app = Rack::Builder.app do
        use RequestReplay::Middleware, hopt
        run lambda{ |en|
          en['PATH_INFO'] = '/bad'
          [200, {}, []]
        }
      end

      app.call(e)
      begin
        sock = serv.accept
        sock.read.should.eq <<-HTTP
GET /?q=1 HTTP/1.1\r
Host: localhost\r
Pork: BEEF\r
\r
      HTTP
      ensure
        sock.close
      end
    end

    should 'rewrite_env' do
      app = Rack::Builder.app do
        use RequestReplay::Middleware, hopt, :rewrite_env => lambda{ |en|
          if en['HTTP_HOST'].start_with?('api.')
            en['PATH_INFO'] = "/api#{en['PATH_INFO']}"
          end
          en
        }, :add_headers => {'Host' => 'eg.com'}
        run lambda{ |_| [200, {}, []] }
      end

      app.call(env.merge('HTTP_HOST' => 'api.localhost'))
      begin
        sock = serv.accept
        sock.read.should.eq <<-HTTP
GET /api/?q=1 HTTP/1.1\r
Host: eg.com\r
Pork: BEEF\r
\r
      HTTP
      ensure
        sock.close
      end
    end
  end
end


test/test_proxy.rb


require 'request-replay/test'

describe RequestReplay::Proxy do
  behaves_like :test

  request = lambda do |env, buf, options={}|
    mock(buf).close
    Thread.new{
      begin
        RequestReplay::Proxy.new(@hopt, options).call(env)
        buf.string
      rescue => e
        p e
      end
    }
  end

  after do
    Muack.verify
  end

  should 'basic' do
    expected = <<-HTTP
GET /?q=1 HTTP/1.1\r
Host: localhost\r
Pork: BEEF\r
Connection: close\r
\r
    HTTP

    buf = StringIO.new
    env = {'rack.hijack' => mock.call{ env['rack.hijack_io'] = buf }.object}.
          merge(@env)

    @verify[request[env, buf], expected]
  end

  should 'add_headers' do
    expected = <<-HTTP
GET /?q=1 HTTP/1.1\r
Host: ex.com\r
Pork: BEEF\r
Connection: close\r
\r
    HTTP

    buf = StringIO.new
    env = {'rack.hijack' => mock.call{ env['rack.hijack_io'] = buf }.object}.
          merge(@env)

    @verify[request[env, buf, :add_headers => {'Host' => 'ex.com'}], expected]
  end

  should 'rewrite_env' do
    expected = <<-HTTP
GET /a?q=1 HTTP/1.1\r
Host: localhost\r
Pork: BEEF\r
Connection: close\r
\r
    HTTP

    buf = StringIO.new
    env = {'rack.hijack' => mock.call{ env['rack.hijack_io'] = buf }.object}.
          merge(@env)
    rewrite_env = lambda{ |env| env['PATH_INFO'] = '/a'; env }
    @verify[request[env, buf, :rewrite_env => rewrite_env], expected]
  end
end


test/test_middleware.rb

require 'request-replay/test'

describe RequestReplay::Middleware do
  behaves_like :test

  should 'PUT' do
    hopt = @hopt
    app = Rack::Builder.app do
      use RequestReplay::Middleware, hopt
      run lambda{ |_| [200, {}, []] }
    end

    app.call(@env.merge('REQUEST_METHOD' => 'PUT'))
    begin
      sock = @serv.accept
      sock.read.should.eq <<-HTTP
PUT /?q=1 HTTP/1.1\r
Host: localhost\r
Pork: BEEF\r
\r
    HTTP
    ensure
      sock.close
    end
  end

  should 'retain original env' do
    hopt = @hopt
    e = @env.dup

    app = Rack::Builder.app do
      use RequestReplay::Middleware, hopt
      run lambda{ |env|
        env['PATH_INFO'] = '/bad'
        [200, {}, []]
      }
    end

    app.call(e)
    begin
      sock = @serv.accept
      sock.read.should.eq <<-HTTP
GET /?q=1 HTTP/1.1\r
Host: localhost\r
Pork: BEEF\r
\r
    HTTP
    ensure
      sock.close
    end
  end

  should 'rewrite_env' do
    hopt = @hopt
    app = Rack::Builder.app do
      use RequestReplay::Middleware, hopt, :rewrite_env => lambda{ |env|
        if env['HTTP_HOST'].start_with?('api.')
          env['PATH_INFO'] = "/api#{env['PATH_INFO']}"
        end
        env
      }, :add_headers => {'Host' => 'eg.com'}
      run lambda{ |_| [200, {}, []] }
    end

    app.call(@env.merge('HTTP_HOST' => 'api.localhost'))
    begin
      sock = @serv.accept
      sock.read.should.eq <<-HTTP
GET /api/?q=1 HTTP/1.1\r
Host: eg.com\r
Pork: BEEF\r
\r
    HTTP
    ensure
      sock.close
    end
  end
end


lib/replay-request.rb

require 'socket'
require 'stringio'

class ReplayRequest
  autoload :Middleware, 'replay-request/middleware'

  NEWLINE      = "\r\n"
  HTTP_VERSION = 'HTTP/1.1'

  def initialize env, host, addhead={}
    @env, (@host, @port), @addhead = env, host.split(':', 2), addhead
    if env['rack.input']
      env['rack.input'].rewind
      @buf = StringIO.new
      IO.copy_stream(env['rack.input'], @buf)
      @buf.rewind
    end
  end

  def start
    write_request
    write_headers
    write_payload
    yield(sock) if block_given?
  ensure
    sock.close
  end

  def write_request
    sock.write("#{request}#{NEWLINE}")
  end

  def write_headers
    sock.write("#{headers}#{NEWLINE}#{NEWLINE}")
  end

  def write_payload
    return unless @buf
    IO.copy_stream(@buf, sock)
  end

  def request
    "#{@env['REQUEST_METHOD']} #{request_path} #{HTTP_VERSION}"
  end

  def headers
    headers_hash.map{ |name, value| "#{name}: #{value}" }.join(NEWLINE)
  end

  def request_path
    "/#{@env['PATH_INFO']}?#{@env['QUERY_STRING']}".
      sub(%r{^//}, '/').sub(/\?$/, '')
  end

  def headers_hash
    @headers_hash ||=
      @env.inject({}){ |r, (k, v)|
        r[capitalize_headers(k[5..-1])] = v if k.start_with?('HTTP_')
        r
      }.merge('Content-Type'   => @env['CONTENT_TYPE'  ],
              'Content-Length' => @env['CONTENT_LENGTH']).
        merge(@addhead).
        select{ |_, v| v }
  end

  def capitalize_headers header
    header.downcase.gsub(/[a-z]+/){ |s| s.capitalize }.tr('_', '-')
  end

  def sock
    @sock ||= TCPSocket.new(@host, Integer(@port || 80))
  end
end


lib/request-replay.rb

require 'socket'
require 'stringio'

class RequestReplay
  autoload :Middleware, 'request-replay/middleware'
  autoload :Proxy     , 'request-replay/proxy'

  NEWLINE        = "\r\n"          .freeze
  HTTP_VERSION   = 'HTTP/1.1'      .freeze
  RACK_INPUT     = 'rack.input'    .freeze
  RACK_ERRORS    = 'rack.errors'   .freeze
  REQUEST_METHOD = 'REQUEST_METHOD'.freeze
  PATH_INFO      = 'PATH_INFO'     .freeze
  QUERY_STRING   = 'QUERY_STRING'  .freeze
  CONTENT_TYPE   = 'CONTENT_TYPE'  .freeze
  CONTENT_LENGTH = 'CONTENT_LENGTH'.freeze
  HEADER_CTYPE   = 'Content-Type'  .freeze
  HEADER_CLENGTH = 'Content-Length'.freeze

  def initialize env, host, options={}
    @env, (@host, @port), @options = env, host.split(':', 2), options
    if env[RACK_INPUT]
      env[RACK_INPUT].rewind
      @buf = StringIO.new
      IO.copy_stream(env[RACK_INPUT], @buf)
      @buf.rewind
      env[RACK_INPUT].rewind
    else
      @buf = nil
    end
  end

  def add_headers
    @options[:add_headers] || {}
  end

  def read_wait
    @options[:read_wait] && Float(@options[:read_wait])
  end

  def start
    write_request
    write_headers
    write_payload
    IO.select([sock], [], [], read_wait) if read_wait
    yield(sock) if block_given?
  rescue => e
    @env[RACK_ERRORS].puts("[#{self.class.name}] Error: #{e.inspect}") if
      @env[RACK_ERRORS]
  ensure
    sock.close
  end

  def write_request
    sock.write("#{request}#{NEWLINE}")
  end

  def write_headers
    sock.write("#{headers}#{NEWLINE}#{NEWLINE}")
  end

  def write_payload
    return unless @buf
    IO.copy_stream(@buf, sock)
  end

  def request
    "#{@env[REQUEST_METHOD] || 'GET'} #{request_path} #{HTTP_VERSION}"
  end

  def headers
    headers_hash.map{ |name, value| "#{name}: #{value}" }.join(NEWLINE)
  end

  def request_path
    "/#{@env[PATH_INFO]}?#{@env[QUERY_STRING]}".
      sub(%r{^//}, '/').sub(/\?$/, '')
  end

  def headers_hash
    @headers_hash ||=
      @env.inject({}){ |r, (k, v)|
        r[capitalize_headers(k[5..-1])] = v if k.start_with?('HTTP_')
        r
      }.merge(HEADER_CTYPE   => @env[CONTENT_TYPE  ],
              HEADER_CLENGTH => @env[CONTENT_LENGTH]).
        merge(add_headers).select{ |_, v| v }
  end

  def capitalize_headers header
    header.downcase.gsub(/[a-z]+/){ |s| s.capitalize }.tr('_', '-')
  end

  def sock
    @sock ||= TCPSocket.new(@host, Integer(@port || 80))
  end
end
